File: solutions/ruby/blackjack/blackjack.rb
module Blackjack
  # Calculate the value of any given card
  def self.parse_card(card)
    case card
    when "ace"
      11
    when "king", "queen", "jack", "ten"
      10
    when "nine"
      9
    when "eight"
      8
    when "seven"
      7
    when "six"
      6
    when "five"
      5
    when "four"
      4
    when "three"
      3
    when "two"
      2
    else
      0
    end
  end

  # Name ranges of values.
  def self.card_range(card1, card2)
    sum = parse_card(card1) + parse_card(card2)
    case
    when sum == 21
      "blackjack"
    when sum >= 17
      "high"
    when sum >= 12
      "mid"
    else
      "low"
    end
  end

  # Implement the decision logic for the first turn.
  def self.first_turn(card1, card2, dealer_card)
    hand = card_range(card1, card2)
    case
    when card1 == "ace" && card2 == "ace"
      "P" # Split
    when hand == "blackjack" && parse_card(dealer_card) < 10
      "W" # auto win
    when hand == "low" || hand == "mid" && parse_card(dealer_card) >= 7
      "H" # Hit
    else
      "S" # Stand
    end
  end
end

File: solutions/go/blackjack/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "ace":
		return 11
	case "two":
		return 2
	case "three":
		return 3
	case "four":
		return 4
	case "five":
		return 5
	case "six":
		return 6
	case "seven":
		return 7
	case "eight":
		return 8
	case "nine":
		return 9
	case "ten":
		return 10
	case "jack":
		return 10
	case "queen":
		return 10
	case "king":
		return 10
	default:
		return 0
	}
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {
	valuePlayer := ParseCard(card1) + ParseCard(card2)
	valueDealer := ParseCard(dealerCard)

	switch {
	case valuePlayer == 22:
		return "P"
	case valuePlayer == 21:
		if valueDealer >= 10 {
			return "S"
		} else {
			return "W"
		}
	case valuePlayer >= 17:
		return "S"
	case valuePlayer >= 12:
		if valueDealer >= 7 {
			return "H"
		} else {
			return "S"
		}
	default:
		return "H"
	}
}
